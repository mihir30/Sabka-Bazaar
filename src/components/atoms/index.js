import Button from "./Button/Button";
import Image from "./Image/Image";
import WithSpinner from "./Spinner/Spinner";

export { Button, Image, WithSpinner };
