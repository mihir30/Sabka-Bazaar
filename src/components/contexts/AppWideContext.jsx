import React, { createContext, useReducer } from "react";
import { reducer } from "../reducers/ReducerContext";

const initialState = {
  count: 0,
  products: {},
};

export const AppWideContext = createContext(initialState);

const AppWideContextProvider = ({ children }) => {
  const [cartItems, dispatch] = useReducer(reducer, initialState);

  return (
    <AppWideContext.Provider value={{ cartItems, dispatch }}>
      {children}
    </AppWideContext.Provider>
  );
};

export default AppWideContextProvider;
