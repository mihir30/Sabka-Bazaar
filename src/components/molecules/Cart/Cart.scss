$background-color-black: rgba(0, 0, 0, 0.4);

.cart {
    width: 2.7rem;
    height: 3rem;
    margin-left: 0rem;
}
.nav-cart-items {
    text-decoration: none;
}
.cart-quantity {
    font-size: medium;
    font-weight: lighter;
    margin: 0 0.5rem;
    text-decoration: none;
}

.btn-cart-button {
    border: none;
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100%;
    width: 100%;
    padding: 0.4rem 0.3rem;
}

.cart-modal-overlay {
    display: none;
}

.cart-modal {
    display: none;
}

@media only screen and (min-width: 1024px) {
    .cart-modal-overlay {
        display: none;
        position: fixed;
        z-index: 1;
        padding-top: 100px;
        left: 0;
        top: 0;
        width: 100%;
        height: 100%;
        overflow: auto;
        background-color: $background-color-black;
    }

    .cart-modal-overlay::-webkit-scrollbar {
        display: none;
    }

    .cart-modal {
        height: 100vh;
        width: 35%;
        float: right;
        display: block;
        margin: auto;
        padding: 20px;
    }
    .btn-cart-button {
        margin: 0.2rem 0 0 0rem;
    }
}
