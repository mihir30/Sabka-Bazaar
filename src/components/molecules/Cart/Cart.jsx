import React, { useEffect, useRef, useState, useContext } from "react";
import { Link } from "react-router-dom";
import { Image, Button } from "../../atoms";
import { CartSegment } from "../../organisms";
import cart from "../../../static/images/cart.svg";
import { AppWideContext } from "../../contexts/AppWideContext";
import "./Cart.scss";
const Cart = () => {
  const cartModalOverlay = useRef(null);
  const [isDesktop, setIsDesktop] = useState(window.innerWidth > 1024);

  const {
    cartItems: { count },
  } = useContext(AppWideContext);
  const countItems = count === 1 ? `${count} item` : `${count} items`;
  const setMedia = () => {
    setIsDesktop(window.innerWidth > 1024);
  };

  useEffect(() => {
    window.addEventListener("resize", setMedia);
    return () => window.removeEventListener("resize", setMedia);
  });

  const handleCartClick = () => {
    if (cartModalOverlay.current.style.display === "none") {
      cartModalOverlay.current.style.display = "block";
    } else {
      cartModalOverlay.current.style.display = "none";
    }
  };

  const handleCloseBtn = () => {
    cartModalOverlay.current.style.display = "none";
  };

  return (
    <>
      {isDesktop ? (
        <>
          <div className="cart-btn">
            <Button
              type="button"
              buttonType="cart-button"
              method={handleCartClick}
            >
              <Image source={cart} alt="cart" className="cart" />
              <span className="cart-quantity" data-testid="count-items">{countItems}</span>
            </Button>
          </div>
          <div className="cart-modal-overlay" ref={cartModalOverlay}>
            <div className="cart-modal">
              <CartSegment closeBtnMethod={handleCloseBtn} />
            </div>
          </div>
        </>
      ) : (
        <>
          <div className="cart-btn">
            <Link to="/cartItems" className="nav-cart-items">
              <Button
                type="button"
                buttonType="cart-button"
                method={handleCartClick}
              >
                <Image source={cart} alt="cart" className="cart" />
                <span className="cart-quantity" data-testid="count-items">{countItems}</span>
              </Button>
            </Link>
          </div>
          <div className="cart-modal-overlay" ref={cartModalOverlay}>
            <div className="cart-modal">
              <CartSegment closeBtnMethod={handleCloseBtn} />
            </div>
          </div>
        </>
      )}
    </>
  );
};

export default Cart;
