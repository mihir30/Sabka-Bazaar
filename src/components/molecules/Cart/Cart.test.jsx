import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { AppWideContext } from '../../contexts/AppWideContext';
import Cart from './Cart';
import { Button } from '../../atoms';

describe("Cart Header Component", () => {
    it('renders the component properly and should display the proper initial count in the cart header component', () => {
        render(<AppWideContext.Provider value={{ cartItems: { count: 0 } }}><BrowserRouter><Cart /></BrowserRouter></AppWideContext.Provider>);
        expect(screen.getByTestId("count-items")).toHaveTextContent('0 items');
    });

    it('should update the count in cart header component', () => {
        const cartCount = 0;
        const addItemToCart = jest.fn();
        render(<Button buttonType="BuyNow"
        children="Buy Now"
        method={() => addItemToCart()}
        data-testid="add-to-cart"></Button>);
        const buttonElement = screen.getByRole("button", { name: /Buy Now/i });
        fireEvent.click(buttonElement);
        render(<AppWideContext.Provider value={{ cartItems: { count: cartCount+1 } }}><BrowserRouter><Cart /></BrowserRouter></AppWideContext.Provider>);
        expect(screen.getByTestId("count-items")).toHaveTextContent('1 item');

    });

    it('should have the right cart header component snapshot', () => {
        const tree = renderer.create(<AppWideContext.Provider value={{ cartItems: { count: 0 } }}><BrowserRouter><Cart /></BrowserRouter></AppWideContext.Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    })

});