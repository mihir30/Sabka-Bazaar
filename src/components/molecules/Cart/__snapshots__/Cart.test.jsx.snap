// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Header Component should have the right cart header component snapshot 1`] = `
Array [
  <div
    className="cart-btn"
  >
    <a
      className="nav-cart-items"
      href="/cartItems"
      onClick={[Function]}
    >
      <button
        className="btn-cart-button"
        onClick={[Function]}
        type="button"
      >
        <img
          alt="cart"
          className="cart"
          data-testid="image"
          src="cart.svg"
        />
        <span
          className="cart-quantity"
          data-testid="count-items"
        >
          0 items
        </span>
      </button>
    </a>
  </div>,
  <div
    className="cart-modal-overlay"
  >
    <div
      className="cart-modal"
    >
      <div
        className="minicart-modal-background"
      >
        <div
          className="minicart-empty"
        >
          <div
            className="cart-quantity-tile"
          >
            <p>
              <strong
                data-testid="my-cart-text"
              >
                My Cart 
              </strong>
            </p>
          </div>
          <div
            className="cart-quantity-tile-modal"
          >
            <p>
              <strong
                data-testid="my-cart-text-modal"
              >
                My Cart 
              </strong>
            </p>
            <i
              className="fas fa-times"
              id="close-btn-empty-cart"
              onClick={[Function]}
            />
          </div>
          <div>
            <p
              className="empty-cart-text"
              data-testid="no-items"
            >
              No items in your cart
            </p>
            <p
              className="empty-cart-desc"
              data-testid="fav-items"
            >
              Your favourite items are just a click away
            </p>
          </div>
          <div
            className="checkout-empty-tile"
          >
            <div>
              <button
                className="btn-checkout"
                onClick={[Function]}
              >
                <span
                  className="start-shopping-btn"
                >
                  <span>
                    Start Shopping
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
